lib/paths.bzl

```diff
@@ -153,6 +153,67 @@ def _normalize(path):
 
     return path or "."
 
+_BASE = 0
+_SEPARATOR = 1
+_DOT = 2
+_DOTDOT = 3
+
+def _is_normalized(str, look_for_same_level_references = True):
+    """Returns true if the passed path contains uplevel references "..".
+
+    Also checks for single-dot references "." if look_for_same_level_references
+    is `True.`
+
+    Args:
+        str: The path string to check.
+        look_for_same_level_references: If True checks if path contains uplevel
+          references ".." or single-dot references ".".
+
+    Returns:
+        True if the path is normalized, False otherwise.
+    """
+    state = _SEPARATOR
+    for c in str.elems():
+        is_separator = False
+        if c == "/":
```

Comment:
Possible incorrect for windows. The native code check is: https://github.com/bazelbuild/bazel/blob/34e3159e342a58b66353cca57ebe0b3d3ad85ff0/src/main/java/com/google/devtools/build/lib/vfs/WindowsOsPathPolicy.java#L231

Reply:
The native check actually uses the host operating system: https://github.com/bazelbuild/bazel/blob/34e3159e342a58b66353cca57ebe0b3d3ad85ff0/src/main/java/com/google/devtools/build/lib/vfs/PathFragment.java#L728

Since we're using `/` throughout the rules, even when they run on Windows, I believe this code is better than the native one. One piece of evidence are also all other functions here, that use just forward slashes.